source/common/http/async_client_impl.cc

```diff
@@ -71,16 +71,16 @@ AsyncClient::Request* AsyncClientImpl::send(RequestMessagePtr&& request,
 }
 
 AsyncClient::Stream* AsyncClientImpl::start(AsyncClient::StreamCallbacks& callbacks,
-                                            const AsyncClient::StreamOptions& options) {
+                                            AsyncClient::StreamOptions& options) {
   std::unique_ptr<AsyncStreamImpl> new_stream{new AsyncStreamImpl(*this, callbacks, options)};
   LinkedList::moveIntoList(std::move(new_stream), active_streams_);
   return active_streams_.front().get();
 }
 
 AsyncStreamImpl::AsyncStreamImpl(AsyncClientImpl& parent, AsyncClient::StreamCallbacks& callbacks,
-                                 const AsyncClient::StreamOptions& options)
+                                 AsyncClient::StreamOptions& options)
     : parent_(parent), stream_callbacks_(callbacks), stream_id_(parent.config_.random_.random()),
-      router_(options.filter_config_ ? options.filter_config_.value().get() : parent.config_,
+      router_(options.filter_config_ ? *options.filter_config_ : parent.config_,
```

Comment:
note: I also tried the opposite approach here, of constructing router with a const config. But that doesn't work because the config isn't really just a read-only configuration; it's got other context bundled into it that requires non-const access, so it was easier to just change the expectations up the stack to deal with non-const configs.

This added 10 files to this PR, but going the other direction (constifying all the way down) appeared to be much harder.

Reply:
I see. It "worked" because std::reference_wrapper behaved like std::unique_ptr by returning constness of reference type regarding of the constness of the wrapper itself. https://en.cppreference.com/w/cpp/utility/functional/reference_wrapper/get

And the new OptRef returns `const T&` if OptRef itself is const. Also its API was inconsistent. Calling OptRef<>::get() was not the same as calling `value().get();` with respect to constness of types.

I think conceptually returning T for cont OptRef is more correct since constness of OptRef<T> should not imply constness of T. Similar to how std::reference_wrapper works.

Reply:
OK I gave that a try; that enabled me to revert the chain of non-constness I had added previously also.

I'm not sure how I feel about extracting a non-const T from a const OptRef<T> -- I guess it's OK but I wasn't thinking about it like that.

Reply:
Interesting stuff, clearly const-ness is tricky